### PR TITLE
Non empty bins

### PR DESCRIPTION
**Pull request by 0x0Louis**

### CODE REVIEW

**Review by Mathieu-Be — APPROVED** [review]

**Review by pecio222 — APPROVED** [review]

File: src/LiquidityHelperContract.sol [hunk 1]
```diff
@@ -208,4 +211,61 @@ contract LiquidityHelperContract {
...
+     * If id is not specified, it will use the active bin id of the pair.
+     * Will check `lengthLeft` non-empty bins on the left and `lengthRight` non-empty bins on the right, so if the user
+     * has liquidity only after the `lengthLeft + 1` bin on the left and `lengthRight + 1` bin on the right, it will return
+     * an empty array.
+     * @param pair The liquidity book pair.
+     * @param user The user.
+     * @param id The specific bin id. (optional)
+     * @param lengthLeft The number of non-empty bins to fetch on the left.
+     * @param lengthRight The number of non-empty bins to fetch on the right.
+     * @return id The bin id used. (id id was not specified, will return the active bin id)
```
Comment (pecio222) [inline src/LiquidityHelperContract.sol]: **if** id;

File: src/LiquidityHelperContract.sol [hunk 2]
```diff
@@ -208,4 +211,62 @@ contract LiquidityHelperContract {
...
+    function getPopulatedBinsReserves(ILBPair pair, uint24 start, uint24 end, uint24 length)
+        external
+        view
+        returns (NonEmptyBinHelper.PopulatedBin[] memory)
+    {
+        return NonEmptyBinHelper.getPopulatedBinsReserves(pair, start, end, length);
+    }
+
+    /**
+     * @notice Fetches the non-empty bins reserves of a liquidity book pair from [start, end] where the user has liquidity.
```
Comment (pecio222) [inline src/LiquidityHelperContract.sol]: [id-lengthLeft,id+lengthRight]

File: src/LiquidityHelperContract.sol [hunk 3]
```diff
@@ -208,4 +211,62 @@ contract LiquidityHelperContract {
...
+     * has liquidity only after the `lengthLeft + 1` bin on the left and `lengthRight + 1` bin on the right, it will return
+     * an empty array.
+     * @param pair The liquidity book pair.
+     * @param user The user.
+     * @param id The specific bin id. (optional)
+     * @param lengthLeft The number of non-empty bins to fetch on the left.
+     * @param lengthRight The number of non-empty bins to fetch on the right.
+     * @return id The bin id used. (id id was not specified, will return the active bin id)
+     * @return The array of populated bins with (id, reserveX, reserveY, shares, totalShares)
+     * The user amounts can be calculated as (shares * reserve{X,Y}) / totalShares. (totalShares > 0)
```
Comment (pecio222) [inline src/LiquidityHelperContract.sol]: totalShares are guaranteed to be > 0 here

File: src/periphery/NonEmptyBinHelper.sol [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+            (uint128 reserveX, uint128 reserveY) = pair.getBin(id);
+            populatedBins[i] = PopulatedBin(id, reserveX, reserveY);
+        }
+
+        return populatedBins;
+    }
+
+    /**
+     * @notice Fetches the non-empty bins reserves of a liquidity book pair from [start, end] where the user has liquidity.
```
Comment (pecio222) [inline src/periphery/NonEmptyBinHelper.sol]: [id-lengthLeft,id+lengthRight]

File: src/periphery/NonEmptyBinHelper.sol [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+     * @param pair The liquidity book pair.
+     * @param user The user.
+     * @param id The specific bin id. (optional)
+     * @param lengthLeft The number of non-empty bins to fetch on the left.
+     * @param lengthRight The number of non-empty bins to fetch on the right.
+     * @return The bin id used. (id id was not specified, will return the active bin id)
+     * @return The array of populated bins with (id, reserveX, reserveY, shares, totalShares)
+     * The user amounts can be calculated as (shares * reserve{X,Y}) / totalShares. (totalShares > 0)
+     */
+    function getBinsReserveOf(ILBPair pair, address user, uint24 id, uint24 lengthLeft, uint24 lengthRight)
```
Comment (pecio222) [inline src/periphery/NonEmptyBinHelper.sol:137]: Might move id as last argument as it's optional? Not sure if it's pattern in solidity , can easily ignore

File: src/periphery/NonEmptyBinHelper.sol [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+     * If id is not specified, it will use the active bin id of the pair.
+     * Will check `lengthLeft` non-empty bins on the left and `lengthRight` non-empty bins on the right, so if the user
+     * has liquidity only after the `lengthLeft + 1` bin on the left and `lengthRight + 1` bin on the right, it will return
+     * an empty array.
+     * @param pair The liquidity book pair.
+     * @param user The user.
+     * @param id The specific bin id. (optional)
+     * @param lengthLeft The number of non-empty bins to fetch on the left.
+     * @param lengthRight The number of non-empty bins to fetch on the right.
+     * @return The bin id used. (id id was not specified, will return the active bin id)
```
Comment (pecio222) [inline src/periphery/NonEmptyBinHelper.sol]: If id

File: src/periphery/NonEmptyBinHelper.sol [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+     * @return The array of populated bins with (id, reserveX, reserveY, shares, totalShares)
+     * The user amounts can be calculated as (shares * reserve{X,Y}) / totalShares. (totalShares > 0)
```
Comment (pecio222) [inline src/periphery/NonEmptyBinHelper.sol]: It's guaranteed for totalShares to be > 0

File: src/LiquidityHelperContract.sol [hunk 8]
```diff
@@ -2,11 +2,14 @@
...
+import {ILBPair} from "joe-v2/interfaces/ILBPair.sol";
+
+import {LiquidityHelper} from "./periphery/LiquidityHelper.sol";
+import {NonEmptyBinHelper} from "./periphery/NonEmptyBinHelper.sol";
 
 /**
- * @title Liquidity Book periphery contract for Liquidity and Fees Amounts
- * @notice Periphery contract to help compute liquidity and fees amounts from amounts and ids.
+ * @title Liquidity Book periphery contract for Liquidity, Fees Amounts and bin fetching.
+ * @notice Periphery contract to help compute liquidity, fees amounts from amounts and ids and fetch bins.
```
Comment (pecio222) [inline src/LiquidityHelperContract.sol:13]: maybe worth adding, that this contract should be called only off-chain, based on quoter experience

**Review by pecio222 — APPROVED** [review]

File: src/periphery/NonEmptyBinHelper.sol (hunk 5 above)
Comment (0x0Louis) [reply src/periphery/NonEmptyBinHelper.sol:137]: I rather like it like that, but I agree both makes sense

File: src/LiquidityHelperContract.sol (hunk 8 above)
Comment (0x0Louis) [reply src/LiquidityHelperContract.sol:13]: Agree lol

**Pull request merged**